### PR TITLE
FuzzySearch: New fuzzy search application

### PR DESCRIPTION
**Pull request by fabianblatz**
This commit adds a fuzzy searching utility in the style of the fzf project: [https://github.com/junegunn/fzf](https://github.com/junegunn/fzf). I designed it as a GUI application for now since we do not yet support an ncurses like library. For other apps it might be interesting to also use the search/scoring algorithms. 

The tool is usefull for any terminal based file manipulation where you want to provide the user with a list of items he can choose from. One workflow I am using in my host machine which also works quite nicely in Serenity is fuzzy opening a file in a text editor:

`te $(fs)`

will first show the user a list of the files in his current working directory and after the users chose one its opened in the text edtior.

The search algorithm is a port of the Go v1 algorithm of fzf, but I also first tested the UI with the levenshtein-distance since it was easier to implement.

Things which can be added over time:
- Highlighting the search term within the result view
- Improving the algorithm/creating new ones
- adding some of the cli options

### Screenshots:

image
image
image

### CODE REVIEW

File: Userland/Applications/FuzzySearch/FuzzySearchAlgorithms.cpp [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+        }
+        ++i;
+    }
+
+    return 1.0 - (double)costs[n] / max(text.length(), pattern.length());
+}
+
+double FuzzySearchAlgorithms::fzf_match_v1(const StringView& text, const StringView& pattern, const SearchOptions& options)
+{
+    //translation of the Go implementation of fzf: https://github.com/junegunn/fzf
```
Comment (linusg) [inline Userland/Applications/FuzzySearch/FuzzySearchAlgorithms.cpp:55]:
```suggestion
    // Translation of the Go implementation of fzf: https://github.com/junegunn/fzf
```

File: Userland/Applications/FuzzySearch/FuzzySearchAlgorithms.cpp [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+        return -1;
+    }
+    size_t pattern_index = 0;
+    Optional<size_t> start_index;
+    Optional<size_t> end_index;
+
+    size_t len_text = text.length();
+    size_t len_pattern = pattern.length();
+
+    auto indexAt = [](int32_t index, int32_t max) {
```
Comment (linusg) [inline Userland/Applications/FuzzySearch/FuzzySearchAlgorithms.cpp:69]:
```suggestion
    auto index_at = [](i32 index, i32 max) {
```

File: Userland/Applications/FuzzySearch/FuzzySearchAlgorithms.cpp [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+{
+    if (!all_of(pattern.begin(), pattern.end(), isascii)) {
+        return -1;
+    }
+    int first_index = 0;
+    int index = 0;
+    for (size_t pattern_index = 0; pattern_index < pattern.length(); pattern_index++) {
+        index = try_skip(text, ignore_case, pattern[pattern_index], index);
+        if (index < 0) {
+
```
Comment (linusg) [inline Userland/Applications/FuzzySearch/FuzzySearchAlgorithms.cpp:185]:
```suggestion
```

**Review by linusg — CHANGES_REQUESTED** [review]
Please use `i32` instead of `int32_t` everywhere. You can also skip the curly braces from if (etc.) statements if the body, including comments, is only one line. The algorithms might be useful elsewhere, maybe those should live in AK or a new LibFuzzySearch instead?

File: Userland/Applications/FuzzySearch/FuzzySearchAlgorithms.cpp [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+#include <string.h>
+
+static constexpr int32_t score_match = 16;
+static constexpr int32_t score_gap_start = -3;
+static constexpr int32_t score_gat_extension = -1;
+
+static constexpr int32_t bonus_boundry = score_match / 2;
+static constexpr int32_t bonus_non_word = score_match / 2;
+static constexpr int32_t bonus_camel123 = bonus_boundry + score_gat_extension;
+static constexpr int32_t bonus_consequtive = -(score_gap_start + score_gat_extension);
```
Comment (Maato) [inline Userland/Applications/FuzzySearch/FuzzySearchAlgorithms.cpp:17]: consecutive

**Comment by bugaevc** [comment]
> `cat | fs`

😢

File: Userland/Applications/FuzzySearch/FuzzySearchAlgorithms.cpp [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+{
+    const size_t m = text.length();
+    const size_t n = pattern.length();
+    if (n == 0)
+        return m;
+    if (m == 0)
+        return n;
+
+    size_t costs[n + 1];
+    memset(&costs, 0, n + 1);
```
Comment (Maato) [inline Userland/Applications/FuzzySearch/FuzzySearchAlgorithms.cpp:30]: Looks like this doesn't zero the whole costs array.

**Comment by awesomekling** [comment]
From the [contribution guidelines](https://github.com/SerenityOS/serenity/blob/7b9051afe516ff889a2d7d1c80aa0e04f9f70c33/CONTRIBUTING.md):

> For your first PR, start with something small to get familiar with the project and its development processes.

Since you have started with something large instead, please at least make an effort to explain what this is, what the use-cases are, why it should be a GUI application and not a command-line application, etc.

**Comment by faxe1008** [comment]
@linusg I resolved all of the points you mentioned besides the tearing out of the algorihms into AK or seperate library, since I do not if they are "good" enough to be a thing of their own. 

@awesomekling This is not my first pull request, accidentally I placed it from another github account, the one which maintains the fork had a (merged) pull request prior to this: [https://github.com/SerenityOS/serenity/commit/cbb06d70146e10bad7157e768669ef1d5f48adde](https://github.com/SerenityOS/serenity/commit/cbb06d70146e10bad7157e768669ef1d5f48adde)
Anyways, will add a small paragraph as you suggested.

File: Userland/Applications/CMakeLists.txt [hunk 6]
```diff
@@ -6,6 +6,7 @@ add_subdirectory(Calendar)
 add_subdirectory(CrashReporter)
 add_subdirectory(Debugger)
 add_subdirectory(DisplaySettings)
+add_subdirectory(FuzzySearch)
```
[the remark this answers is not in the document]
Comment (fabianblatz) [reply Userland/Applications/CMakeLists.txt]: Fixed it.

File: Userland/Applications/FuzzySearch/FuzzySearchAlgorithms.cpp (hunk 1 above)
Comment (fabianblatz) [reply Userland/Applications/FuzzySearch/FuzzySearchAlgorithms.cpp]: Fixed it.

File: Userland/Applications/FuzzySearch/FuzzySearchAlgorithms.cpp (hunk 2 above)
Comment (fabianblatz) [reply Userland/Applications/FuzzySearch/FuzzySearchAlgorithms.cpp]: Fixed it.

File: Userland/Applications/FuzzySearch/FuzzySearchAlgorithms.cpp (hunk 3 above)
Comment (fabianblatz) [reply Userland/Applications/FuzzySearch/FuzzySearchAlgorithms.cpp]: Fixed it.

File: Userland/Applications/FuzzySearch/FuzzySearchAlgorithms.cpp (hunk 4 above)
Comment (fabianblatz) [reply Userland/Applications/FuzzySearch/FuzzySearchAlgorithms.cpp]: Yup, removed the typo

File: Userland/Applications/FuzzySearch/FuzzySearchAlgorithms.cpp (hunk 5 above)
Comment (fabianblatz) [reply Userland/Applications/FuzzySearch/FuzzySearchAlgorithms.cpp]: Yep, did not take the type into account, changed it to a `sizeof` usage.

**Comment by faxe1008** [comment]
> > `cat | fs`
> 
> 😢

The original tool did not support directly opening a file, it only handled stdin so I made my version akin to it. I guess adding it as an optional positional argument might be interesting.

**Comment by bugaevc** [comment]
> The original tool did not support directly opening a file, it only handled stdin so I made my version akin to it. I guess adding it as an optional positional argument might be interesting.

Oh, that's not what I mean at all. Reading from stdin is the right thing to do; but `cat foo | fs` is another example of Useless Use of Cat :crying_cat_face:

I've seen somebody do `dd if=somefile | less` once...

**Comment by faxe1008** [comment]
@linusg I split the fuzzy matching into AK as you suggested.

**Pull request closed without merging**